blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,951 @@
+        mpool_locked.add(&smsg_vec[1]).unwrap();
+        mpool_locked.add(&smsg_vec[2]).unwrap();
+        mpool_locked.add(&smsg_vec[3]).unwrap();
+
+        mpool_locked
+            .api
+            .lock()
+            .unwrap()
+            .set_state_sequence(&sender, 0);
+        mpool_locked
+            .head_change(Vec::new(), vec![Tipset::new(vec![a]).unwrap()])
+            .unwrap();
+        assert_eq!(mpool_locked.get_sequence(&sender).unwrap(), 4);
+
+        mpool_locked
+            .api
+            .lock()
+            .unwrap()
+            .set_state_sequence(&sender, 1);
+        mpool_locked
+            .head_change(Vec::new(), vec![Tipset::new(vec![b.clone()]).unwrap()])
+            .unwrap();
+        assert_eq!(mpool_locked.get_sequence(&sender).unwrap(), 4);
+
+        mpool_locked
+            .api
+            .lock()
+            .unwrap()
+            .set_state_sequence(&sender, 0);
+        mpool_locked
+            .head_change(vec![Tipset::new(vec![b]).unwrap()], Vec::new())
+            .unwrap();
+        assert_eq!(mpool_locked.get_sequence(&sender).unwrap(), 4);
+
+        let (p, _) = mpool_locked.pending().unwrap();
+        assert_eq!(p.len(), 3);
+    }
+
+    #[test]
+    fn test_async_message_pool() {
+        let keystore = MemKeyStore::new();
+        let mut wallet = Wallet::new(keystore);
+        let sender = wallet.generate_addr(SignatureType::Secp256k1).unwrap();
+        let target = wallet.generate_addr(SignatureType::Secp256k1).unwrap();
+
+        let mut tma = TestApi::new();
+        tma.set_state_sequence(&sender, 0);
+
+        let mpool = MessagePool::new(tma, "mptest".to_string()).unwrap();
+        let mut mpool_locked = mpool.lock().unwrap();
+
+        let mut smsg_vec = Vec::new();
+        for i in 0..3 {
+            let msg = create_smsg(&target, &sender, wallet.borrow_mut(), i);
+            smsg_vec.push(msg);
+        }
+
+        assert_eq!(mpool_locked.get_sequence(&sender).unwrap(), 0);
+        mpool_locked.push(&smsg_vec[0]).unwrap();
+        assert_eq!(mpool_locked.get_sequence(&sender).unwrap(), 1);
+        mpool_locked.push(&smsg_vec[1]).unwrap();
+        assert_eq!(mpool_locked.get_sequence(&sender).unwrap(), 2);
+        mpool_locked.push(&smsg_vec[2]).unwrap();
+        assert_eq!(mpool_locked.get_sequence(&sender).unwrap(), 3);
+
+        drop(mpool_locked);
+
+        let header = mock_block(1, 1);
+        let tipset = Tipset::new(vec![header.clone()]).unwrap();
+
+        let updater = mpool.lock().unwrap();
+        let temp = updater.api.clone();
+        let mut api = temp.lock().unwrap();
+        drop(updater);
+
+        let ts = tipset.clone();
+        task::block_on(async move {
+            // updater.api.lock().unwrap().set_block_messages(&header, vec![message]);
+            api.set_heaviest_tipset(Arc::new(ts)).await;
+        });
+
+        // sleep allows for async block to update mpool's cur_tipset
+        sleep(Duration::new(2, 0));
```

Comment:
since youre calling block_on, why do you need to sleep? set_heaviest_tipset should be guaranteed to run first no?

Reply:
Actually no, so what happens is that set_heaviest_tipset will send a subscriber notification that the head has changed, so then the async block in the MP initializer will get notified so it will run the head_change method.

Reply:
So technically set_heaviest_tipset will run first, but there isn't a guarantee that the head change will run first before the check below